File: euler27.py
# Найдите произведение коэффициентов a < 1000 и b <= 1000 для
# квадратичного выражения, которое выражает максимальное число
# простых чисел для последовательных значений n, начиная с n = 0


from datetime import datetime
from functools import lru_cache

n = 1_000


def eratosfen(number):
    """Решето Эратосфена: генерация простых чисел вплоть до заданного числа"""
    sieve = list(range(number + 1))
    sieve[1] = 0
    for i in sieve:
        if i > 1:
            for j in range(2 * i, len(sieve), i):
                sieve[j] = 0
    return sieve


@lru_cache(2 ** 5)
def isprime(number: int):
    """Проверка числа на простоту перебором делителей"""
    if number in {2, 3, 5, 7}: return True
    if number < 2 or number % 2 == 0: return False
    if number % 3 == 0 or number % 5 == 0: return False
    a, k = number, 0
    for i in range(5, int(number**0.5), 6):
        k += 1 if a % i == 0 or a % (i + 2) == 0 else 0
    return True if k <= 0 else False

# 1 вариант
start_time = datetime.now()
primes_num = sorted(set(eratosfen(n)))
primes_num[0], primes, max_n_primes = 1, primes_num[:], 0
for b in primes_num:
    for a in primes_num:
        i = 0
        for x in range(a, -2 * a, -2 * a):
            while True:
                quadratic = i**2 + x * i + b
                if quadratic not in primes:
                    if isprime(quadratic) and quadratic > 0:
                        primes.append(quadratic)
                    else:
                        if i - 1 > max_n_primes:
                            max_n_primes, axb = i - 1, x * b
                        break
                i += 1
print(axb)
print(datetime.now() - start_time)

# 2 вариант
start_time = datetime.now()
pass
print(datetime.now() - start_time)

File: go/euler27.go
// euler27 находит произведение коэффициентов a и b для квадратичного
// выражения с максимальным числом простых чисел.
//
// Решение задачи Эйлера №27.
package main

import (
	"fmt"
	"math"
	"time"
)

func eratosfen(number int) []bool {
	sieve := make([]bool, number+1)
	for i := range sieve {
		sieve[i] = true
	}
	sieve[0] = false
	sieve[1] = false
	for i := 2; i*i <= number; i++ {
		if sieve[i] {
			for j := i * i; j <= number; j += i {
				sieve[j] = false
			}
		}
	}
	return sieve
}

func isPrime(number int) bool {
	if number < 2 {
		return false
	}
	if number == 2 || number == 3 || number == 5 || number == 7 {
		return true
	}
	if number%2 == 0 || number%3 == 0 || number%5 == 0 {
		return false
	}
	sqrt := int(math.Sqrt(float64(number)))
	for i := 5; i <= sqrt; i += 6 {
		if number%i == 0 || number%(i+2) == 0 {
			return false
		}
	}
	return true
}

func euler27() {
	n := 1000

	startTime := time.Now()
	primesNum := make([]int, 0)
	sieve := eratosfen(n)
	for i := 2; i <= n; i++ {
		if sieve[i] {
			primesNum = append(primesNum, i)
		}
	}
	primesNum = append([]int{1}, primesNum...)
	maxNPrimes := 0
	axb := 0
	for _, b := range primesNum {
		for _, a := range primesNum {
			i := 0
			for x := a; x > -2*a; x -= 2 * a {
				quadratic := i*i + x*i + b
				if !isPrime(quadratic) || quadratic <= 0 {
					if i-1 > maxNPrimes {
						maxNPrimes = i - 1
						axb = x * b
					}
					break
				}
				i++
			}
		}
	}
	fmt.Println(axb)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	fmt.Println(time.Since(startTime))
}

func main() {
	euler27()
}
